src/SpellsReference/Data/Repositories/SpellRepository.cs

```diff
@@ -38,6 +48,31 @@ public List<Spell> List()
             return _context.Spells.ToList();
         }
 
+        public List<Spell> List(SpellFilterViewModel filter)
+        {
+            var spells = _context.Spells
+                .Where(s =>
+                    (!filter.Level.HasValue || filter.Level.Value == s.Level) &&
+                    (!filter.School.HasValue || filter.School.Value == s.School)
+                    )
```

Comment:
If you haven't already, you'll want to verify that this `Where` operator expression can be converted to SQL by EF. When EF can't convert the expression to SQL, it'll retrieve all of the data from the database and filter using LINQ to Objects.

Reply:
👍